zqd/storage/zng.go

```diff
@@ -0,0 +1,246 @@
+	}
+	return s, s.readInfoFile()
+}
+
+func (s *ZngStorage) join(args ...string) string {
+	args = append([]string{s.path}, args...)
+	return filepath.Join(args...)
+}
+
+func (s *ZngStorage) Open(span nano.Span) (zbuf.ReadCloser, error) {
+	zctx := resolver.NewContext()
+	f, err := os.Open(s.join(allZngFile))
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+
+		// Couldn't read all.zng, check for an old space with all.bzng
+		bzngFile := strings.TrimSuffix(allZngFile, filepath.Ext(allZngFile)) + ".bzng"
+		f, err = os.Open(s.join(bzngFile))
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return nil, err
+			}
+			r := zngio.NewReader(strings.NewReader(""), zctx)
+			return zbuf.NopReadCloser(r), nil
+		}
+	}
+	// XXX plz keep track of how long this takes
+	return s.index.NewReader(f, zctx, span)
+}
+
+type spanWriter struct {
+	span nano.Span
+}
+
+func (w *spanWriter) Write(rec *zng.Record) error {
+	if rec.Ts == 0 {
+		return nil
+	}
+	first := w.span == nano.Span{}
+	s := nano.Span{Ts: rec.Ts, Dur: 1}
+	if first {
+		w.span = s
+	} else {
+		w.span = w.span.Union(s)
+	}
+	return nil
+}
+
+func (s *ZngStorage) Rewrite(ctx context.Context, zr zbuf.Reader) error {
+	if !s.wsem.TryAcquire(1) {
+		return zqe.E(zqe.Conflict, ErrWriteInProgress)
+	}
+	defer s.wsem.Release(1)
+
+	// For the time being, this endpoint will overwrite any underlying data.
+	// In order to get rid errors on any concurrent searches on this space,
+	// write zng to a temp file and rename on successful conversion.
+	tmppath := s.join(allZngTmpFile)
+	zngfile, err := os.Create(tmppath)
+	if err != nil {
+		return err
+	}
+
+	fileWriter := zngio.NewWriter(zngfile, zio.WriterFlags{StreamRecordsMax: s.streamsize})
+	spanWriter := &spanWriter{}
+	zw := zbuf.MultiWriter(fileWriter, spanWriter)
+
+	if err := s.write(ctx, zw, zr); err != nil {
+		zngfile.Close()
+		os.RemoveAll(tmppath)
+		return err
+	}
+	if err := fileWriter.Flush(); err != nil {
+		return err
+	}
+
+	if err := zngfile.Close(); err != nil {
+		os.RemoveAll(tmppath)
+		return err
+	}
+	if err := os.Rename(tmppath, s.join(allZngFile)); err != nil {
+		return err
+	}
+	return s.SetSpan(spanWriter.span)
+}
+
+func (s *ZngStorage) write(ctx context.Context, zw zbuf.Writer, zr zbuf.Reader) error {
+	out, err := driver.Compile(ctx, zngWriteProc, zr, false, nano.MaxSpan, zap.NewNop())
+	if err != nil {
+		return err
+	}
+	d := &zngdriver{zw}
+	return driver.Run(out, d, nil)
+}
+
+// Clear wipes all data from storage. Will wait for any ongoing write operations
+// are complete before doing this.
+func (s *ZngStorage) Clear(ctx context.Context) error {
+	if err := s.wsem.Acquire(ctx, 1); err != nil {
+		return err
+	}
+	defer s.wsem.Release(1)
+	if err := os.Remove(s.join(allZngFile)); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return s.UnsetSpan()
+}
+
+// XXX This is not thread safe and it should be.
+func (s *ZngStorage) SetSpan(span nano.Span) error {
```

Comment:
The behavior for this function doesn't match the name, and in pcap ingest, and this behavior requires the `UnsetSpan` , `SetSpan` sequence during pcap ingest. Can the unioning be removed so this function just sets the space's span?

Reply:
The unioning is needed because otherwise spans from the pcap ingests snapshots would overwrite the the full span size we get from generating the pcap index. I think the best long term solution would be to have space.Span() take the union of PcapStorage.Span and LogStorage.Span, but I put this in for now to arrive at feature parity. I should add a comment mentioning this here.

Reply:
Then for now, I suggest you change the name from `SetSpan` to `UnionSpan`, as the behavior is surprising given the name.

Reply:
Agreed, much clearer.